### PR TITLE
Add new abstraction to help with form validation, and use it within the "Extract Columns" modal

### PR DESCRIPTION
**Pull request by seancolsen**
## Why I opened this PR

So I got a little carried away on a somewhat tangential front end project during the past few work days, and now I want to come clean and introduce it. I know we ought avoid tangents like this right now, being laser focused on the live demo and release, so I'm definitely guilty of following my curiosity too far down the rabbit hole here. Sorry :grimacing:  But I'm hoping it will pay off...

Last week I was working on building the UI for the record summary template configuration. It's a relatively simple form, but I was frustrated with having done the same thing over and over again to implement client-side form validation. And many times I simply _haven't_ implemented validation because doing so is kind of tedious.

I had an idea for a small helper utility, and when I started chasing that idea it blossomed into something that I think is really cool!

## What's in this PR

- 6d7bb94bc94ca9d15f4f2508b25af2b8cd235a7c adds the the form validation system, and makes a few tweaks to our component library so the form system can better utilize it.
- 15cc7a01deb507a42a07eae6feb0734da8715ac4 refactors the "Move Columns" modal to use the new system, improving the client-side validation quite a bit.

## Why I didn't choose an existing form validation library

I actually spent quite a bit of time looking at existing solutions and took some [notes](https://hackmd.io/YPvOIu7rSHKx6eOB9UnIYA) on my research. I can review them with anyone who is interested. I spent some time playing with some of these libraries and didn't like any of them.

## An example form

Let's look at the form in my tests...

```ts
type Beverage = 'juice' | 'water' | 'beer';

const name = requiredField('', [uniqueWith(['Jim'], 'Name already taken.')]);
const age = requiredField<number | undefined>(undefined, [
  min(16, 'You must be at least 16 to register.'),
]);
const beverage = requiredField<undefined | Beverage>(undefined);

const form = makeForm({ name, age, beverage }, [
  comboInvalidIf(
    [age, beverage],
    ([a, b]) => b === 'beer' && a < 21,
    'Too young for beer.',
  ),
]);
```

```html
<Field label="Name" field={name} />
<Field label="Age" field={age} input={NumberInput} />
<Field label="Beverage" field={beverage} />
<FormSubmit {form} onCancel={form.reset} onProceed={save} initiallyHidden />
```

image

image

image

image

image

## How it works

1. You make a new field:

    ```ts
    const name = requiredField('', [uniqueWith(['Jim'], 'Name already taken.')]);
    ```

1. `name` is now a `Field`, which mostly looks like this:

    ```ts
    interface Field<T> extends Writable<T> {
      isRequired: boolean;
      initialValue: T;
      isDirty: Readable<boolean>;
      fieldErrors: Readable<string[]>;
      comboErrors: Readable<string[]>;
      errors: Readable<string[]>;
      isValid: Readable<boolean>;
      showsError: Readable<boolean>;
      reset: () => void;
    }
    ```

1. You can read/write/bind `$name` to get its value, and you can read its validation errors from the inner derived store too.

1. You assemble one or more fields into a form like so:

    ```ts
    const form = makeForm({ name, age, beverage }, [
      comboInvalidIf(
        [age, beverage],
        ([a, b]) => b === 'beer' && a < 21,
        'Too young for beer.',
      ),
    ]);
    ```

1. `form` is now a `Form`, which mostly looks like this:

    ```ts
    interface FormContent {
      values: 'All the field values will be here, fully typed';
      isValid: boolean;
      isDirty: boolean;
    }
    interface Form extends Readable<FormContent> {
      reset: () => void;
    }
    ```

1. You can read `$form.isValid` to check if all the validation errors have passed. The `FormSubmit` helper component renders a `CancelOrProceedButtonPair` that can only proceed when the entire form is valid.

1. We supplied this "combo validator" when creating the form:

    ```ts
    comboInvalidIf(
      [age, beverage],
      ([a, b]) => b === 'beer' && a < 21,
      'Too young for beer.',
    )
    ```

    The form will set that error message on each field used within the validator if the validator fails. And the validation function is fully type-safe by inferring the types from the fields.

1. `Field.svelte` is a relatively simple component that binds the field store to an input, displaying error messages as needed.

    ```html
    <Field label="Name" field={name} />
    ```

    By default Field.svelte will bind to a `TextInput`.

1. The Move Columns modal has a more fully-featured demo of `Field.svelte`

    ```html
    <Field
      field={columns}
      input={{ component: SelectProcessedColumns, props: { availableColumns } }}
      label="Columns to Move"
      layout="stacked"
    >
      <span slot="help">
        These columns will be removed from the current table and moved to the
        linked table.
      </span>
    </Field>
    ```

    Here we have specified the input component and given it props too. I even managed to get this working in a type-safe way.

## Where the code lives

- The form validation system is currently at `src/components/form`. I'd like to eventually extract it into its own package, separate from Mathesar _and_ separate from the component library, but we can talk about that later.

### Before

The form basically had no client-side validation

image

### After

Adding complex validation rules is much less tedious now!

image

## Going forward

- If we can adopt this approach, I think we can continue to refine it as needed. I have some more ideas.
- I'd like to use this form validation system to finish my work on the Record Summary template config form.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by pavish — APPROVED** [review]
Looks good to me! Nice work @seancolsen 

* We use the function name `makeForm` in both Form-builder and here. It would be good to maybe rename one of them, but I think it's okay to keep the same name for now.
* We could probably leverage this within Form-builder in the future. We have a small validation logic within form-builder, which is defined by a json because we envisioned it to be serializable and stored on the backend at some point. With all the changes we're going through, we could give up on serialization and keep everthing uniform. If such a need to serialize forms arise in the future, we can find a way to serialize this library.

I'll leave it to @rajatvijay to review before merging this in.

**Comment by rajatvijay** [comment]
@seancolsen this looks really good. I have some ideas around extending this but I'll add it to the cooldown meeting on coming tuesday.

**Review by rajatvijay — APPROVED** [review]

**Pull request merged**
